File: daily/Stack/155-Min-stack.go
package daily

type MinStack struct {
	stack    []int
	minStack []int
}

func MinStackConstructor() MinStack {
	return MinStack{}
}

func (ms *MinStack) Push(val int) {
	ms.stack = append(ms.stack, val)
	if len(ms.minStack) == 0 || val <= ms.minStack[len(ms.minStack)-1] {
		ms.minStack = append(ms.minStack, val)
	}
}

func (ms *MinStack) Pop() {
	top := ms.stack[len(ms.stack)-1]
	ms.stack = ms.stack[:len(ms.stack)-1]
	if top == ms.minStack[len(ms.minStack)-1] {
		ms.minStack = ms.minStack[:len(ms.minStack)-1]
	}
}

func (ms *MinStack) Top() int {
	return ms.stack[len(ms.stack)-1]
}

func (ms *MinStack) GetMin() int {
	return ms.minStack[len(ms.minStack)-1]
}

File: daily/Stack/155-Min-stack.py
class MinStack:
    def __init__(self):
        self.stacks = []
        self.min_stacks = []

    def push(self, value: int) -> None:
        self.stacks.append(value)
        if not self.min_stacks or value <= self.min_stacks[-1]:
            self.min_stacks.append(value)

    def pop(self) -> None:
        topElement = self.stacks.pop()
        if topElement == self.min_stacks[-1]:
            self.min_stacks.pop()

    def top(self) -> int:
        return self.stacks[-1]

    def getMin(self) -> int:
        return self.min_stacks[-1]
